### PR TITLE
Add Reaper Control Plane Deployment Mode

### LINKED ISSUE #1277: New Reaper deployment mode: Control Plane
We want to allow deploying a single instance of Reaper which will repair all K8ssandraClusters managed by a single control plane.
This requires to create a new deployment mode in Reaper (aside from PER_DC and SINGLE), which we can name "CONTROL_PLANE".
A Reaper custom resource would need to be created separately prior to the K8ssandraCluster objects creation, with this mode.
Any K8ssandraCluster object without Reaper settings would be registered by the operator into the control plane Reaper and repairs would be managed by it.

We would disallow enabling Reaper in clusters if the control plane instance exists to keep things simple in the first iteration.
This means either you have one centralized Reaper managing all clusters or you have individual Reaper instances per cluster.

```[tasklist]
### Definition of Done
- [x] the Reaper API has a new deployment mode named `CONTROL_PLANE`
- [x] A Reaper running with memory storage, the depoyment mode must be `CONTROL_PLANE` or `SINGLE`
- [x] A `CONTROL_PLANE` reaper comes up without contacting any K8ssandra clusters (for storage or registration)
- [x] A `SINGLE` reaper won't deploy more than once
- [x] The k8ssandraCluster object has a field in the reaper spec that defines which Reaper to register with
- [ ] ~In the K8ssandraCluster spec, a reaperRef is added to point to the Reaper instance that will handle the repairs~
- [x] When a CP Reaper instance exists, k8ssandra-operator registers all K8ssandraClusters in it
- [ ] ~If a K8ssandraCluster object gets created with Reaper settings while a control plane Reaper exists, the K8ssandraCluster objects should override control plane Reaper with their own defined Reaper settings.~
```

### PR DESCRIPTION
**Pull request by rzvoncek**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #1277 

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: config/webhook/kustomization.yaml [hunk 1]
```diff
@@ -14,6 +14,9 @@ patchesJson6902:
     name: k8ssandra-operator-validating-webhook-configuration
     kind: ValidatingWebhookConfiguration
   patch: |-
+    - op: replace
```
Comment (rzvoncek) [inline config/webhook/kustomization.yaml:17]: I don't remember if this is something auto-generated, or I did this by accident and should rever this.

Comment (burmanm) [reply config/webhook/kustomization.yaml:17]: No, I added this for you.

**Review by adejanovski — CHANGES_REQUESTED** [review]

**Comment by Miles-Garnsey** [comment]
Notes:

- @rzvoncek is having some issues with the password/username not being found when trying to register the external reaper cluster inside the k8ssandraCluster. 
- We've discussed defaulting the ReaperRef.namespace to the same namespace as the K8ssandraCluster itself when trying to retrieve dependant resources in the controller. That should fix this.

File: apis/reaper/v1alpha1/reaper_types.go [hunk 2]
```diff
@@ -75,6 +75,11 @@ type ReaperTemplate struct {
 	// +optional
 	UiUserSecretRef *corev1.LocalObjectReference `json:"uiUserSecretRef,omitempty"`
 
+	// When there is a CONTROL_PLANE Reaper out there, this field allows registering a K8ssandra cluster to it.
+	// Populating this field disables some operator behaviour related to setting Reaper up.
+	// +optional
+	ReaperRef corev1.ObjectReference `json:"reaperRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply apis/reaper/v1alpha1/reaper_types.go]: I've moved the `ReaperRef` to `ReaperClusterTemplate`. 
We still need to tell if _we are reconciling a CP reaper_, based on just `ReaperTemplate`, but that is now done via a check for `DatacenterRef` presence.

File: controllers/reaper/reaper_controller.go [hunk 3]
```diff
@@ -275,6 +297,29 @@ func (r *ReaperReconciler) reconcileDeployment(
...
+		return *actual.Spec.Replicas
+	case *appsv1.StatefulSet:
+		return *actual.Spec.Replicas
+	default:
+		logger.Error(fmt.Errorf("unexpected type %T", actualDeployment), "Failed to get deployment replicas")
+		return math.MaxInt32
+	}
+}
+
+func setDeploymentReplicas(desiredDeployment *client.Object, singleReplica *int32, logger logr.Logger) {
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/reaper/reaper_controller.go]: Pushed a commit doing this.

File: controllers/k8ssandra/reaper_test.go [hunk 4]
```diff
@@ -278,3 +270,61 @@ func verifyReaperSecretAnnotationAdded(t *testing.T, f *framework.Framework, ctx
...
+				Resources: corev1.VolumeResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: resource.MustParse("1Gi"),
+					},
+				},
+			},
+			AutoScheduling: reaperapi.AutoScheduling{
+				Enabled: true,
+				// we need to be explicit to prevent checking the cass-dc spec, which is not available if reaper's in control plane mode
+				RepairType: "ADAPTIVE",
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/reaper_test.go:381]: pushed a commit resolving this, including a test. but I hope I understood @adejanovski 's thought correctly.

File: controllers/k8ssandra/reaper.go [hunk 5]
```diff
@@ -256,3 +260,27 @@ func getSingleReaperDcName(kc *api.K8ssandraCluster) string {
...
+
+func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+) result.ReconcileResult {
+	manager := reaper.NewManager()
+	manager.SetK8sClient(r)
+	if username, password, err := manager.GetUiCredentials(ctx, kc.Spec.Reaper.UiUserSecretRef, kc.Namespace); err != nil {
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/reaper.go:276]: I've fixed this. Had to a) not create the reaper's CQL secret and b) do not put it into cass-dc CRD.

File: controllers/k8ssandra/k8ssandracluster_controller_test.go [hunk 6]
```diff
@@ -81,6 +99,12 @@ func TestK8ssandraCluster(t *testing.T) {
 			ManagementApi:    managementApiFactory,
 			Recorder:         mgr.GetEventRecorderFor("k8ssandracluster-controller"),
 		}).SetupWithManager(mgr, clusters)
+		err = (&reaper.ReaperReconciler{
+			ReconcilerConfig: config.InitConfig(),
+			Client:           mgr.GetClient(),
+			Scheme:           mgr.GetScheme(),
+			NewManager:       newMockManager,
+		}).SetupWithManager(mgr)
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/k8ssandracluster_controller_test.go]: The webhooks are gone for now, will come back via https://github.com/riptano/mission-control/issues/1453

**Pull request closed without merging**